File: apple/RecKit/Sources/RecKit/Jobs/JobRunner.swift
import Foundation
import Network
import os

/// docs/10 "잡 상태 머신", as much of it as the executor wiring has an opinion about.
public enum JobRunStatus: Sendable {
    case pending
    case running
    case waiting
    case done
    case failed
    case needsAuth
    /// docs/10 "Drive 용량 초과": parked until the user frees space and asks again. Like every other
    /// parked state it is not something [NextRun] comes back for — the core does not poll Drive.
    case needsSpace
    case skippedShort
}

/// One row of the queue, reduced to what deciding *when to come back* needs.
public struct JobRunSnapshot: Sendable, Equatable {
    public let status: JobRunStatus
    public let nextRunAt: Date?

    public init(status: JobRunStatus, nextRunAt: Date? = nil) {
        self.status = status
        self.nextRunAt = nextRunAt
    }
}

/// What one call of `ReclyCore.runDueJobs()` made of itself.
public struct JobPass: Sendable {
    /// The core's executor was already inside a pass, so this one saw nothing at all.
    public let alreadyRunning: Bool
    public let ran: Int

    public init(alreadyRunning: Bool, ran: Int) {
        self.alreadyRunning = alreadyRunning
        self.ran = ran
    }
}

/// Everything [JobRunner] needs from the core, and so the whole of what a test has to stand in for
/// (the Android twin is `JobFacade`). A runner that reached for `ReclyCore` directly could only be
/// tested against a database.
public protocol JobQueue: AnyObject, Sendable {
    func now() -> Date
    func runDueJobs() async throws -> JobPass
    /// The queue as it stands after a pass — read to arm the successor.
    func jobs() async throws -> [JobRunSnapshot]
}

/// When the queue has to be looked at again (docs/11 A5 deliverable 3, docs/12 "실행기"). Pure on
/// purpose: it is the one piece of the executor wiring with arithmetic in it.
public enum NextRun {
    /// The earliest moment a job could next make progress: a `PENDING` job is due immediately, a
    /// `WAITING` one when its backoff elapses — including a backoff that already elapsed while a
    /// long upload held the pass, which is why past instants are kept rather than filtered out.
    /// Everything else (`DONE`, `FAILED`, `NEEDS_AUTH`, `NEEDS_SPACE`, `SKIPPED_SHORT`) waits for a
    /// person.
    ///
    /// `RUNNING` is deliberately not here: a row left running by a kill is recovered by the next
    /// pass, and the five-minute timer brings one. Arming a zero-delay timer for it would wake the
    /// machine to do the same nothing.
    public static func at(_ jobs: [JobRunSnapshot], now: Date) -> Date? {
        jobs.compactMap { job -> Date? in
            switch job.status {
            case .pending: return now
            // A WAITING row with no instant is a lost write, not a job to strand: treat it as due.
            case .waiting: return job.nextRunAt ?? now
            default: return nil
            }
        }.min()
    }

    /// Never negative — an overdue job means "now". nil is "there is nothing to come back for".
    public static func delay(_ jobs: [JobRunSnapshot], now: Date) -> TimeInterval? {
        at(jobs, now: now).map { max(0, $0.timeIntervalSince(now)) }
    }
}

/// docs/12 "실행기": the app process is the only thing that runs the queue on a Mac, and it calls
/// `runDueJobs()` on four triggers — a job just enqueued, the five-minute timer, the network coming
/// back, and the successor armed from the queue itself. The phone shares it (docs/13 I3) and adds
/// a fifth of its own — the app becoming active — through [jobsDue]'s sibling [run].
///
/// Two invariants, ported from the phone's `WorkflowWorker` (docs/11 A5):
///
/// 1. **Every pass leaves a successor behind it.** The queue is only ever woken by an event or a
///    timer, so a pass that ends without arming the next one is a queue that stalls. The successor
///    is recomputed from the job table after the pass ([NextRun]), with [generation] catching
///    whatever landed while the pass was running. A pass never cancels the successor: a queue that
///    reads empty may simply have been read too early.
/// 2. **A pass that found the core busy reshapes nothing.** It saw no queue of its own, so all it
///    guarantees is that *someone* comes back — shortly after the pass in flight should be done.
@MainActor
public final class JobRunner {
    /// docs/12 "실행기": (b), the standing five-minute timer.
    public static let interval: TimeInterval = 5 * 60
    /// How long after a pass that found the core already busy to come back. Long enough that the
    /// pass in flight has usually finished and armed its own successor (which replaces this one),
    /// short enough that a pass killed mid-flight is picked up again.
    public static let followUp: TimeInterval = 60

    private let queue: JobQueue
    /// The queue as the last completed pass left it — the menu reads `NEEDS_AUTH` off this.
    private let onPass: ([JobRunSnapshot]) -> Void
    /// How the successor is armed, when it is not this runner's own one-shot `Timer`. A test hands
    /// in its own so that nothing actually fires and the delays can be asserted on.
    private let armOverride: ((TimeInterval) -> Void)?

    /// Ticks once per "something became runnable". It is why a pass can tell "the queue is
    /// genuinely empty" from "the queue was empty when I looked". Monotonic and never reset: a
    /// pass only ever compares its own before-and-after.
    ///
    /// Its own lock rather than the main actor's isolation: the signal is raised wherever the
    /// enqueue happened — the core answers on its own dispatcher — and a hop would put it *after*
    /// the comparison it exists to lose the race to.
    private let generation = Generation()

    private let logger = Logger(subsystem: CoreBridge.appName, category: "jobs")
    private var ticker: Timer?
    private var successor: Timer?
    private var monitor: NWPathMonitor?
    /// The last path state seen, so only the *return* of the network runs a pass.
    private var online = true

    public init(
        queue: JobQueue,
        arm: ((TimeInterval) -> Void)? = nil,
        onPass: @escaping ([JobRunSnapshot]) -> Void = { _ in }
    ) {
        self.queue = queue
        self.onPass = onPass
        self.armOverride = arm
    }

    /// Triggers (b) and (c). Not in `init` so a test can build a runner without a timer or a path
    /// monitor attached to it.
    public func start() {
        let ticker = Timer(timeInterval: Self.interval, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.run() }
        }
        RunLoop.main.add(ticker, forMode: .common)
        self.ticker = ticker

        let monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { [weak self] path in
            let satisfied = path.status == .satisfied
            Task { @MainActor in self?.networkChanged(satisfied) }
        }
        monitor.start(queue: DispatchQueue.global(qos: .utility))
        self.monitor = monitor
        run()
    }

    /// Trigger (a): a stop that queued a job, a retry, a sign-in that unparked one. The signal goes
    /// first, so a pass already in flight can tell that its own view of the queue is stale even if
    /// the job landed after it read the table.
    @discardableResult
    public func jobsDue() -> Task<Void, Never> {
        signalDue()
        return run()
    }

    /// Ticks the counter without running a pass. Separate from [jobsDue] because that ordering is
    /// the contract: the signal has to be visible before the pass that might miss the job.
    public nonisolated func signalDue() {
        generation.tick()
    }

    /// One pass, and whatever it decides to arm behind it. The task is returned so a test can wait
    /// for it; nothing in the app does.
    @discardableResult
    public func run() -> Task<Void, Never> {
        Task { await pass() }
    }

    private func pass() async {
        // Snapshotted before the pass and compared after the queue is read, never in between: a
        // job enqueued while this pass runs lands in neither, and the counter is the only evidence
        // left that it happened.
        let generation = self.generation.current
        do {
            let summary = try await queue.runDueJobs()
            if summary.alreadyRunning {
                let delay = signalled(generation) ? 0 : Self.followUp
                arm(delay)
                logger.info("job.pass.busy followUpSec=\(Int(delay), privacy: .public)")
                return
            }
            let jobs = try await queue.jobs()
            // The compare has to come after the read. A signal that lands while the queue is being
            // read is exactly the race, and comparing first would look right and miss it.
            let delay = signalled(generation) ? 0 : NextRun.delay(jobs, now: queue.now())
            // Nothing to arm is not the same as cancelling what is there: a stale successor costs
            // one pass that finds nothing and arms nothing in turn, which terminates.
            if let delay { arm(delay) }
            onPass(jobs)
            logger.info(
                """
                job.pass ran=\(summary.ran, privacy: .public) queued=\(jobs.count, privacy: .public) \
                nextSec=\(delay.map { String(Int($0)) } ?? "-", privacy: .public)
                """
            )
        } catch {
            // The pass itself could not happen. The five-minute timer is the retry; arming a
            // successor off a queue we failed to read would be arming it off nothing.
            logger.error("job.pass.failed error=\(String(describing: error), privacy: .private)")
        }
    }

    /// True when something became runnable after this pass started, so the queue read is stale.
    private func signalled(_ generation: Int) -> Bool { self.generation.current != generation }

    private func networkChanged(_ satisfied: Bool) {
        let returned = satisfied && !online
        online = satisfied
        guard returned else { return }
        logger.info("job.network.back")
        run()
    }

    /// [JobRunner.generation], with the lock that lets any thread raise it.
    private final class Generation: @unchecked Sendable {
        private let lock = NSLock()
        private var value = 0

        var current: Int { lock.withLock { value } }

        func tick() { lock.withLock { value += 1 } }
    }

    /// Exactly one successor, always replaced. There is deliberately no way to cancel it: a pass
    /// that finds an empty queue arms nothing and leaves whatever is standing alone.
    private func arm(_ delay: TimeInterval) {
        if let armOverride {
            armOverride(delay)
            return
        }
        successor?.invalidate()
        let timer = Timer(timeInterval: max(delay, 0), repeats: false) { [weak self] _ in
            Task { @MainActor in self?.run() }
        }
        RunLoop.main.add(timer, forMode: .common)
        successor = timer
    }
}

File: windows/app/src/main/kotlin/app/recly/windows/jobs/JobRunner.kt
@file:OptIn(ExperimentalTime::class)

package app.recly.windows.jobs

import java.net.NetworkInterface
import java.util.concurrent.atomic.AtomicLong
import kotlin.coroutines.cancellation.CancellationException
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds
import kotlin.time.ExperimentalTime
import kotlin.time.Instant
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import recly.core.ReclyCore
import recly.core.job.Job
import recly.core.job.RunSummary
import recly.core.platform.Logger

/**
 * Everything [JobRunner] needs from the core, and so the whole of what a test has to stand in for
 * (the phone's twin is `JobFacade`, the Mac's is `JobQueue`). A runner that reached for
 * [ReclyCore] directly could only be tested against a database.
 */
interface JobQueue {
    fun now(): Instant

    suspend fun runDueJobs(): RunSummary

    /** The queue as it stands after a pass — read to arm the successor. */
    suspend fun jobs(): List<Job>
}

/** The real one. Every core call is hopped onto `CoreDeps.io` (docs/10 "동시성 · 스레딩"). */
class CoreJobQueue(private val core: ReclyCore) : JobQueue {
    override fun now(): Instant = core.deps.clock.now()

    override suspend fun runDueJobs(): RunSummary = withContext(core.deps.io) { core.runDueJobs() }

    override suspend fun jobs(): List<Job> = withContext(core.deps.io) { core.jobs.list() }
}

/**
 * docs/14 "실행기": the app process is the only thing that runs the queue on Windows, and it calls
 * `runDueJobs()` on four triggers — a job just enqueued, the five-minute timer, the network coming
 * back, and the successor armed from the queue itself.
 *
 * Two invariants, the same two the phone's `WorkflowWorker` and the Mac's `JobRunner` have:
 *
 * 1. **Every pass leaves a successor behind it.** The queue is only ever woken by an event or a
 *    timer, so a pass that ends without arming the next one is a queue that stalls. The successor
 *    is recomputed from the job table after the pass ([NextRun]), with [generation] catching
 *    whatever landed while the pass was running. A pass never cancels the successor: a queue that
 *    reads empty may simply have been read too early.
 * 2. **A pass that found the core busy reshapes nothing.** It saw no queue of its own, so all it
 *    guarantees is that *someone* comes back — shortly after the pass in flight should be done.
 */
class JobRunner(
    private val queue: JobQueue,
    private val scope: CoroutineScope,
    private val logger: Logger,
    /**
     * How the successor is armed, when it is not this runner's own delayed coroutine. A test hands
     * in its own so that nothing actually fires and the delays can be asserted on.
     */
    private val arm: ((Duration) -> Unit)? = null,
    /** The queue as the last completed pass left it — the tray reads `NEEDS_AUTH` off this. */
    private val onPass: (List<Job>) -> Unit = {},
) {
    /**
     * Ticks once per "something became runnable". It is why a pass can tell "the queue is genuinely
     * empty" from "the queue was empty when I looked". Monotonic and never reset: a pass only ever
     * compares its own before-and-after.
     */
    private val generation = AtomicLong()

    private val lock = Any()
    private var successor: kotlinx.coroutines.Job? = null

    /** The last network state seen, so only the *return* of it runs a pass. */
    private var online = true

    /** Triggers (b) and (c), plus a pass now: a job left parked by the last run is due. */
    fun start() {
        scope.launch {
            while (isActive) {
                pass()
                delay(INTERVAL)
            }
        }
        scope.launch {
            online = hasNetwork()
            while (isActive) {
                delay(NETWORK_POLL)
                val now = hasNetwork()
                val returned = now && !online
                online = now
                if (returned) {
                    logger.log(Logger.Level.INFO, "job.network.back")
                    pass()
                }
            }
        }
    }

    /**
     * Trigger (a): a stop that queued a job, a retry, a sign-in that unparked one. The signal goes
     * first, so a pass already in flight can tell that its own view of the queue is stale even if
     * the job landed after it read the table.
     */
    fun jobsDue() {
        signalDue()
        scope.launch { pass() }
    }

    /** Ticks the counter without running a pass — the ordering above is the contract. */
    fun signalDue() {
        generation.incrementAndGet()
    }

    /** One pass, and whatever it decides to arm behind it. */
    suspend fun pass() {
        // Snapshotted before the pass and compared after the queue is read, never in between: a job
        // enqueued while this pass runs lands in neither, and the counter is the only evidence left
        // that it happened.
        val generation = this.generation.get()
        try {
            val summary = queue.runDueJobs()
            if (summary.alreadyRunning) {
                val delay = if (signalled(generation)) Duration.ZERO else FOLLOW_UP
                armNext(delay)
                logger.log(Logger.Level.INFO, "job.pass.busy", mapOf("followUpSec" to delay.inWholeSeconds))
                return
            }
            val jobs = queue.jobs()
            // The compare has to come after the read. A signal that lands while the queue is being
            // read is exactly the race, and comparing first would look right and miss it.
            val delay = if (signalled(generation)) Duration.ZERO else NextRun.delay(jobs, queue.now())
            // Nothing to arm is not the same as cancelling what is there: a stale successor costs
            // one pass that finds nothing and arms nothing in turn, which terminates.
            delay?.let { armNext(it) }
            onPass(jobs)
            logger.log(
                Logger.Level.INFO,
                "job.pass",
                mapOf(
                    "ran" to summary.jobIds.size,
                    "queued" to jobs.size,
                    "nextSec" to (delay?.inWholeSeconds?.toString() ?: "-"),
                ),
            )
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            // The pass itself could not happen. The five-minute timer is the retry; arming a
            // successor off a queue we failed to read would be arming it off nothing.
            logger.log(Logger.Level.ERROR, "job.pass.failed", error = e)
        }
    }

    /** True when something became runnable after this pass started, so the queue read is stale. */
    private fun signalled(generation: Long): Boolean = this.generation.get() != generation

    /**
     * Exactly one *pending* successor, always replaced. There is deliberately no way to cancel it: a
     * pass that finds an empty queue arms nothing and leaves whatever is standing alone.
     *
     * The successor lets go of its own reference the moment it stops being a timer and becomes a
     * pass. Without that, a pass that finds the core busy would arm through this and cancel the
     * coroutine it is itself running in — killing the step in flight and leaving its row RUNNING
     * for the next pass to recover. The phone's scheduler has the same rule: `rec-jobs` is KEEP,
     * never REPLACE, because cancelling work in flight spends a retry (docs/11 A5).
     */
    private fun armNext(after: Duration) {
        arm?.let {
            it(after)
            return
        }
        synchronized(lock) {
            successor?.cancel()
            successor = scope.launch {
                delay(after)
                val running = coroutineContext.job
                synchronized(lock) { if (successor === running) successor = null }
                pass()
            }
        }
    }

    /**
     * docs/14 "실행기" (c). There is no `NWPathMonitor` on the JVM, so it is polled: an interface
     * that is up, not loopback and has an address is as much as `java.net` can say about whether
     * this machine is on a network. A false positive costs one pass that fails and retries.
     */
    private suspend fun hasNetwork(): Boolean = withContext(Dispatchers.IO) {
        runCatching {
            NetworkInterface.networkInterfaces().anyMatch { candidate ->
                candidate.isUp && !candidate.isLoopback && candidate.inetAddresses().findAny().isPresent
            }
        }.getOrDefault(true)
    }

    companion object {
        /** docs/14 "실행기": (b), the standing five-minute timer. */
        val INTERVAL: Duration = 5.minutes

        /**
         * How long after a pass that found the core already busy to come back. Long enough that the
         * pass in flight has usually finished and armed its own successor (which replaces this
         * one), short enough that a pass killed mid-flight is picked up again.
         */
        val FOLLOW_UP: Duration = 60.seconds

        internal val NETWORK_POLL: Duration = 30.seconds
    }
}
